Bulk Grades API (v1 grades api w/courses endpoint)

kabirkhan opened the pull request:
### Provides functionality for a course staff member to request grades for all students enrolled in a course, not just a single user. 

Uses a 'v1' naming convention:

Possible routes are:
**anyone can get their own grade still (same as v0)**
  GET /api/grades/v1/course_grade/{course_id}/users/?username={username}  

**course staff member gets grades for specific student (same as v0)**
  GET /api/grades/v1/course_grade/{course_id}/users/?username={username}  

**course staff member gets grades for all students enrolled in course (new for v1)**
  GET /api/grades/v1/course_grade/{course_id}/all_users/

edx-webhook commented:
Thanks for the pull request, @kabirkhan! I've created [OSPR-2272](https://openedx.atlassian.net/browse/OSPR-2272) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

nedbat commented:
jenkins ok to test

nedbat commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,352 @@
+                course.id, serialize=False
+            )
+        except Exception as ex:
+            print(ex)
```
Please remove this.

nedbat commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,352 @@
+                error_code='user_does_not_exist'
+            )
+
+    def _make_grade_response(self, user, course, course_grade, use_email=None):
```
It seems odd to use None for a boolean parameter.

nedbat commented:
(I commented on two things I happened to notice. We'll get an in-depth review from someone else.)

nedbat commented:
@nasthagiri @edx/educator-product Thoughts?

nasthagiri commented:
Thank you for opening the PR.  I have a couple of questions/comments on the high-level approach:

1. The current [/course_grade/../users endpoint](https://github.com/edx/edx-platform/blob/a28d87ea7d3061de5be751f737c4f02217d7fa9c/lms/djangoapps/grades/api/views.py#L112) should already support returning a paginated list of grades.  However, it was limited to filtering by a single `username`, per recommendation for "Explicit Filters" in the [edX REST API Conventions](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/18350757/edX+REST+API+Conventions#edXRESTAPIConventions-2.IdentifyingResources).  I would expect this PR to extend the existing endpoint so grades for all enrollments in the course are returned.  For example, the URL would be: 
```
GET `/api/grades/v1/course_grade/{course_id}/users/?all_users=True`
```
Is there a reason why this approach wasn't taken?

2. At a high level, the proposed modifications to the Grades API are additive and do not introduce any breaking changes to the existing version.  Given that, I believe we can make these changes to the existing V0 of the API.  Is there a reason that a new major version of the API is proposed? 

3. Having said that, I believe the existing V0 can be bumped up to V1 assuming your changes can be made in an additive fashion to the current V0.  If we do this, the new version can be exposed at the routing level (in [urls.py](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/grades/api/urls.py)) and shouldn't require duplicating the code.

kabirkhan commented:
Originally the implementation was split into two different views for future work in the area around security. However, we are changing that future implementation to not require this. So I'll make the architecture updates you suggested.

nasthagiri commented:
@kabirkhan Ah. That makes sense.  Thanks for the historical perspective.

kabirkhan commented:
@nasthagiri Do we still want this to be a configurable addition? Using the ENABLE_BULK_GRADES_API flag

mettursathish commented:
With enabling the use and enforcement of OAuth Scopes in the open edX platform coming soon, my preference is to differentiate single and bulk user as two different endpoints. Validation of scopes becomes easier and manageable with **different endpoint** and not based on the request parameter passed.

Example: 
Single | GET  /api/grades/v0/course_grade/{course_id}/users/?username={username} 	| Scope: Read
Bulk 	| GET   /api/grades/v0/course_grade/{course_id}/all_users				        | Scope: Statistics

With your recommended of single endpoint implementation, the validation of scope becomes a conditional and may not be the best way; also hard to modify in the future. Thoughts?

My 2 cents: Also, if we opt to go with two endpoints, my recommendation is to move to v1. I agree, there are no breaking changes, but its improving performance with the introduction to the new endpoint(bulk).

nasthagiri commented:
@kabirkhan I don't believe we need a config setting (ENABLE_BULK_GRADES_API) for the bulk API work.  This PR enables the use of that API for staff users only, which is limited.  And we will eventually limit use of it by external apps using scopes.  We will want a config setting for the OAuth Scopes work.

nasthagiri commented:
@mettursathish Perhaps we need to have a meeting to flesh this out since it seems we are not on the same page.
Per our design proposal at [Include Organization in Scopes|https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/oauth_dispatch/docs/decisions/0007-include-organizations-in-tokens.rst#scopes-examples], we are proposing there is only a single scope for reading grades.  End users do not need to distinguish between bulk access and single access - so there's no need for separate scopes for these.

Regarding version number, we increase the version number only if there is a breaking interface change.  Performance improvements are implementation details and don't need to affect the consumer of the API.

mettursathish commented:
In our initial design [PR](https://github.com/Microsoft/edx-platform/blob/96f60cfbdd4621eaf20daf1ff5318b561ffde622/lms/envs/common.py) we added 'grades:statistics' to Retrieve grades for all students in a course.

Approach to continue to have a decorator with staff access, then we are good. If the API relies on the scope, how to differentiate scope between a learner request and restricted OAuth request. 

Since the permission depends on the scopes, I was going with a new scope statistics. I’ll setup a call tomorrow(03/07) and we can go from there. Its good that we are getting clarity now.

mettursathish commented:
Based on the discussion today(03/07), below is the proposed 
v0 - /api/grades/v0/course_grade/{course_id}/users/?username={username}

v1 (two different endpoints for single and bulk data) 
Single -  /api/grades/v1/course_grade/{course_id}/users/{username} 	
Bulk    -  /api/grades/v1/course_grade/{course_id}/users/			

We will be discussing tomorrow regarding the proposal.

nasthagiri commented:
@mettursathish @kabirkhan After thinking about this some more, reviewing our [API conventions](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/18350757/edX+REST+API+Conventions), brainstorming with others (@ormsbee, @BenjiLee), looking at other existing similar edX APIs (Courses, Completion, etc), reviewing the [URI RFC for path/query usage](https://tools.ietf.org/html/rfc3986#section-3.4), and reviewing an [industry example](https://api.stackexchange.com/docs), here's what I'm thinking:

## API URIs
Note that we may not support all variations for performance reasons.  But this is to show what a possible set of `GET` calls can look like.

### Grades Courses Resource
This is based on our current API conventions and how other existing edX APIs are designed.
```
/grades/v1/courses/              Courses grades for all users for all courses
  ?course_id={}&username={}        Course grade for filtered user(s) in filtered course(s)
  ?course_id={}                    Course grades for all users in filtered course(s)
  ?username={}                     Course grades for filtered user(s) in all their courses

/grades/v1/courses/{course_id}  Course grades for all users for a course
  ?username={}                     Course grade for filtered user(s) in a course 
```

### Grades Users Resource
This is based on the desired need for a user-specific endpoint. We don't currently use this design pattern elsewhere in our system (other than for the user API itself). However, I do see us possibly supporting this pattern **(in the future)** for certain APIs to allow users and apps to easily get user-specific data without using a filtering API.
```
/grades/v1/users/me/courses        Course grades for authenticated user in all her courses

/grades/v1/users/{user_id}/courses Course grades for a user in all her courses
  ?course_id={}                      Course grades for a user in filtered course(s)
```

### Grades Policy Resource
This is similar to the v0 Grades policy API, but follows the URI RFC by putting the `course_id` resource under a proper hierarchical container (`courses`).
```
/grades/v1/policy/courses/            Grading policy for all courses
  ?course_id={}                         Grading policy for filtered course(s)

/grades/v1/policy/courses/{course_id} Grading policy for a course
```

## Scopes
Note that the first part of the scope is separated by `:` to denote a hierarchically **namespaced** scope, while the second part of the scope uses `;` to denote the **filters** that are required to be honored by the implementing endpoint.
 
### Client Credentials
When a trusted application makes server-to-server calls.  The service user and organization affiliation is embedded in the JWT token.
```
grades:read              For a highly privileged application that can access any grade
grades:read;content_org  For an application to access grades only affiliated with content provided by its organization
grades:read;user_org;    For an application to access grades only affiliated with users provided by its organization
```

### Authorization Code
When a user-approved application makes client-to-server calls on behalf of the approving user.  The user's information is embedded in the JWT token.
```
grades:read;content_org;me For an application to access the token user's grades for only content provided by its organization
grades:read;user_org;me    For an application to access the user's grades (per enterprise use case)
```

### Password-based
When a trusted Mobile app makes calls on behalf of the user.  The user's information is embedded in the JWT token.  This is not an immediate use case since mobile apps don't get Grades and don't use JWT authentication at this time.
```
grades:read;me  For trusted mobile application to access any grade of the token user
```

Please take a look and we can determine how to proceed.

nasthagiri commented:
On another note, @ormsbee who keeps a continual watch on our performance status is curious to understand the exact use cases that are pushing Microsoft to implement the Bulk API.  Do you have a write-up of this somewhere?  In particular, he is asking whether a real-time hook that gets triggered whenever a grade changes would work for your needs?

nasthagiri commented:
FYI @doctoryes @feanil Please review above comments.

doctoryes commented:
I reviewed the comments. At first, the absence of an org filter to denote a privileged application set off my alarms - I would have expected an `admin` filter or the like. But on second thought, I think no filter indicating privileged access should be fine.

nasthagiri commented:
@e0d FYI on https://github.com/edx/edx-platform/pull/17602#issuecomment-372064092

kabirkhan commented:
We are discussing this implementation and design. Will respond back with our intended direction next week.

mettursathish commented:
After discussion with the internal and edX team, the plan is not to change the scopes but to apply  filters (content organization) based on the Authorization Grant. Scope Validation will happen at the gateway, which is at the rest framework authentication layer. The API layer needs to apply the filter and respond to the request accordingly.

Scope | Allowed Access
-- | --
certificates:read | Retrieve any certificate
grades:read | Retrieve any grade
enrollment:read | Retrieve any enrollment

Based on the Authentication Grant, the JWT token created will have filter parameters accordingly. 

## Client Credentials: 
When a trusted application makes server-to-server calls. The service user and organization affiliation is embedded in the JWT token.  

## Example:
``` 
{
  "scopes": ["grades:read"],
  "filters":  ["content_org:Microsoft"],
  "version": "1.0",
  "aud": "lms-key",
  "iss": "http://127.0.0.1:8000/oauth2",
  "preferred_username": "staff",
  "exp": 1520874306,
  "iat": 1520838306,
  "sub": "e5de420d61dbb35b3097bb25b7a9b59a"
} 
```

## Authorization Code & Password-based: 

When a user-approved application makes client-to-server calls on behalf of the approving user(Authorization Code) [or] When a trusted Mobile app makes calls on behalf of the user(Password-based). The user filter “me” will be embedded in the JWT token.

## Example:
```
{
  "scopes": ["grades:read"],
  "filters":  [ "content_org:Microsoft", "user:me" ],
  "version": "1.0",
  "aud": "lms-key",
  "iss": "http://127.0.0.1:8000/oauth2",
  "preferred_username": "staff",
  "exp": 1520874306,
  "iat": 1520838306,
  "sub": "e5de420d61dbb35b3097bb25b7a9b59a"
} 
```

nedbat commented:
@kabirkhan @mettursathish Where does this work stand now? Is the ball back in your court for implementation?

mettursathish commented:
Yes the ball is on our court. Once its ready, we will keep you posted.

kabirkhan commented:
Any guidance on the failing tests? The failures seem to be unrelated to these changes.

kabirkhan commented:
@nasthagiri @nedbat

nedbat commented:
jenkins run all

nasthagiri commented:
@kabirkhan The quality failures are related to the PR.  See https://build.testeng.edx.org/job/edx-platform-quality-flow-pr/2540/Diff_Quality_Report/

I also suggest you rebase with master if you have other flaky failures.

nedbat commented:
@kabirkhan I believe you will need to rebase this onto master to get the tests running properly again.

mettursathish commented:
All checks have passed!!

nasthagiri commented on common/djangoapps/enrollment/data.py:
```diff
@@ -93,6 +93,31 @@ def get_course_enrollment(username, course_id):
+    """Based on the course id, return all user enrollments in the course
+    Args:
+        course_id (str): The course to retrieve enrollment information for.
+        serialize (bool): Boolean denoting whether to serialize the enrollments
```
Per YAGNI, no need to add a `serialize` parameter.
Also, we don't want to have an API with a default parameter value that can potentially cause a bad performance issue - returning _all_ enrollments within a course.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,233 @@
+        else:
+            access_action = 'staff'
+
+        course = self._get_course(course_id, request.user, access_action)
```
This endpoint doesn't need to get access to the `course` object - so let's avoid calling the Mongo database to retrieve the course.
Instead, deserialize the `course_id` to a `CourseKey` object and pass the `course_id` to the `CourseGradeFactory` calls below.

kabirkhan replied:
@nasthagiri How should I query the existence of this course without looking at the modulestore? If I'm hitting this api, I'd like to receive a 404 course does not exist error if I try to get data for a course that doesn't exist but still has a valid course key.

How should I handle this case?

nasthagiri replied:
@kabirkhan You can call [CourseOverview.get_from_id](https://github.com/edx/edx-platform/blob/648b13dc4783205ccb9022cdfd451ed77fde7a07/openedx/core/djangoapps/content/course_overviews/models.py#L268).

kabirkhan replied:
But that still calls the modulestore if the CourseOverview object doesn't exist.

kabirkhan replied:
@nasthagiri

nasthagiri replied:
Good point.  I meant [CourseOverview.get_from_ids_if_exists](https://github.com/edx/edx-platform/blob/062b04aee732513cd5fbc6acde27d8d93e1113e3/openedx/core/djangoapps/content/course_overviews/models.py#L307).  It doesn't have a great interface since it returns a `dict`.  So feel free to create a new method on `CourseOverview` for `get_from_id_if_exists` to retrieve a single course if it exists.

kabirkhan replied:
Ok, I already went ahead and added the new method, glad you agree. I'll submit my changes in a few hours.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,233 @@
+        Return:
+            A JSON serialized representation of the requesting user's current grade status.
+        """
+        use_email = str_to_bool(request.GET.get('use_email'))
```
Can we remove `use_email` as a boolean parameter?
Either always return the email address or add the email address if it's specified in a [requested_fields](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/18350757/edX+REST+API+Conventions#edXRESTAPIConventions-4.URLParameters) parameter.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,233 @@
+                status_code=status.HTTP_403_FORBIDDEN,
+                developer_message='The user requested does not match the logged in user.',
+                error_code='user_mismatch'
+            )
```
Instead of the above code from L72-92, use the [IsUserInUrlOrStaff Mixin](https://github.com/edx/edx-platform/blob/e8469a45aaddfa83fe671978cb6230744b2524c3/openedx/core/lib/api/permissions.py#L128).

kabirkhan replied:
This mixin doesn't seem to apply to course staff roles. The current implementation allows course or global staff to request the course. Is this not intended? Should it just be global staff?

nasthagiri replied:
I think it's better to be consistent with the behavior of the other endpoints right now, which, yes, do require global staff access.
In the subsequent PR this will be improved once we use scopes.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,233 @@
+
+        if username:
+            grade_user = self._get_effective_user(request, course)
+            if isinstance(grade_user, Response):
```
You can remove this check once `IsUserInUrlOrStaff mixin` is used - as mentioned above.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,233 @@
+            )
+
+            if not enrollments_in_course:
+                return self.make_error_response(
```
Why are no enrollments considered a `404`?  It should just be an empty list that's returned, right?

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,233 @@
+            grades = CourseGradeFactory().iter(users, course)
+
+            response = []
+            for user, course_grade, __ in grades:
```
How do you plan to handle errors returned in the iterator?  The 3rd parameter is an error string that can be returned in the response.

kabirkhan replied:
What is your recommendation for handling errors here? It doesn't feel like the call should fail if there is one exception thrown but all other users are graded successfully. Should it just be a silent except since the `CourseGradeFactory.iter` method logs the exception?

nasthagiri replied:
Agree.  We don't want the entire call to fail because of a single user.
Also, this used to be more of an issue before when we computed grades on the fly and didn't persist them.  At this time, errors for individual users are less likely since we're just reading their grade from the database.  However, given the current API for `CourseGradeFactory.iter`, we need to handle that case.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,233 @@
+        return {
+            'user': user,
+            'course_key': str(course.id),
+            'passed': course_grade.passed,
```
This will generate a `NullPointer` exception when `course_grade` is `None` in case of error strings returned by `CourseGradeFactory.iter`.  See comment below.

kabirkhan commented:
@nasthagiri I agree with the rest of your feedback. Just need to fix up the tests a bit. I'll push my fixes by tomorrow.

kabirkhan commented:
@nasthagiri All tests are passing now. I believe I've addressed all your feedback.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+
+    def perform_authentication(self, request):
+        """
+        Ensures that the user is authenticated (e.g. not an AnonymousUser), unless DEBUG mode is enabled.
```
Where is `DEBUG` mode checked?  Is this an outdated comment?

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+        * Get course grades if all user who are enrolled in a course.
+        The currently logged-in user may request all enrolled user's grades information.
+    **Example Request**
+        GET /api/grades/v1/courses/{course_id}/   - Get grades for all users in course
```
Please include both example requests - including the one where `course_id` is passed as a filter via a `GET` parameter.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+class CourseGradesView(GradeViewMixin, GenericAPIView):
+    """
+    **Use Case**
+        * Get course grades if all user who are enrolled in a course.
```
if all user -> of all users

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+    """
+    **Use Case**
+        * Get course grades if all user who are enrolled in a course.
+        The currently logged-in user may request all enrolled user's grades information.
```
Nit: Add "if they are allowed."

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+        GET /api/grades/v1/courses/{course_id}/   - Get grades for all users in course
+    **GET Parameters**
+        A GET request may include the following parameters.
+        * course_id: (required) A string representation of a Course ID.
```
Add `username` as an additional possible `GET` parameter.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+        If the request for information about the course grade
+        is successful, an HTTP 200 "OK" response is returned.
+        The HTTP 200 response has the following values.
+        * username: A string representation of a user's username passed in the request.
```
Also, add "email" as another returned value in the response.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+        * username: A string representation of a user's username passed in the request.
+        * course_id: A string representation of a Course ID.
+        * passed: Boolean representing whether the course has been
+                  passed according the course's grading policy.
```
the -> to the

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+        * passed: Boolean representing whether the course has been
+                  passed according the course's grading policy.
+        * percent: A float representing the overall grade for the course
+        * letter_grade: A letter grade as defined in grading_policy (e.g. 'A' 'B' 'C' for 6.002x) or None
```
Nit: grading_policy -> grading policy
(without the underscore)

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+        Gets a course progress status.
+        Args:
+            request (Request): Django request object.
+            course_id (string): URI element specifying the course location.
```
Add "Can also be passed as a `GET` parameter instead."

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+
+        if username:
+            grade_user = self._get_effective_user(request, course_key)
+            if isinstance(grade_user, Response):
```
Nit: Interspersing code to explicitly check for error handling is not ideal from a Code readability standpoint - especially since we can use `Exceptions` for exceptional control flow.  It would be ideal if we can have the helper methods raise exceptions instead of returning `Response` codes.  However, this is not a merge blocker.

kabirkhan replied:
I fully agree, I adapted the code organization from the v0 grades API to make the transition easier.

nasthagiri commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,192 @@
+            return Response([
+                self._make_grade_response(grade_user, course_key, course_grade)
+            ])
+        else:
```
Nit: For improved code readability, it would be ideal if the code in the `if` clause and the code in the `else` clause were in their own smaller helper methods.  That would keep this main method smaller - and easier to read.

nasthagiri commented on openedx/core/djangoapps/content/course_overviews/models.py:
```diff
@@ -324,6 +324,29 @@ def get_from_ids_if_exists(cls, course_ids):
+    def get_from_id_if_exists(cls, course_id):
+        """
+        Return a CourseOverview for the provided course_id if it exists.
+        Returns None if no CourseOverview exists with the provided course_id
```
Why choose to return `None` instead of just raising the exception?

kabirkhan replied:
From a dev perspective I'd expect `get_from_id` to return an exception if it doesn't exist but `get_from_id_if_exists` to me indicates that the method will catch the exception and just return None.

nasthagiri replied:
ok.

kabirkhan commented on common/djangoapps/enrollment/data.py:
```diff
@@ -93,6 +93,22 @@ def get_course_enrollment(username, course_id):
         return None
 
 
+def get_user_enrollments(course_id):
```
I was trying to be consistent with this file. Should I fix the other methods here?

Could have a method that checks if the `course_key` param is not already a `CourseLocator` instance then convert using `CourseKey.from_string`. 

It feels confusing to make this method different from the others in the file.

nasthagiri replied:
This module is using the following convention:
* `course_id` (string): for `string` serialized form of a `CourseKey` object
* `course_key` (CourseKey): for an instance of a `CourseKey` object

As @AdeyemiA said, I was pointing out the mismatch between what this method's docstring says and what the code says.

Yes, in order to remain consistent with the rest of the methods, sure, keep with the `course_id` name.  Given that, just update the docstring and not the code.  Thanks.

kabirkhan commented on lms/djangoapps/grades/api/v1/urls.py:
```diff
@@ -0,0 +1,29 @@
+        r'^courses/{course_id}/$'.format(
+            course_id=settings.COURSE_ID_PATTERN,
+        ),
+        views.CourseGradesView.as_view(), name='course_grades'
```
Django `reverse` accept `args` or `kwargs` but not both. So I believe this should work fine.

nasthagiri replied:
ok.

AdeyemiA commented on common/djangoapps/enrollment/data.py:
```diff
@@ -93,6 +93,22 @@ def get_course_enrollment(username, course_id):
+def get_user_enrollments(course_id):
+    """Based on the course id, return all user enrollments in the course
+    Args:
+        course_key (CourseKey): Identifier of the course
```
The parameter, 'course_id', does not match the documentation, 'course_key'.

nasthagiri commented:
@kabirkhan We're on the last mile now.  ETA on final changes?

kabirkhan commented:
@nasthagiri I've addressed your feedback last week. I don't understand why the jenkins quality build is failing. There's nothing in the diff-quality report.

nasthagiri commented:
@kabirkhan It looks like the quality error threshold has exceeded.  Search for `lms/djangoapps/grades/api/tests/test_views.py` in https://build.testeng.edx.org/job/edx-platform-quality-flow-pr/3620/artifact/test_root/log/run_quality.out.log and remove those errors.  In case you can't access that page, here are quality errors in grades/api:

```
************* Module lms.djangoapps.grades.api.views
lms/djangoapps/grades/api/views.py:222: [W0221(arguments-differ), CourseGradingPolicy.get] Parameters differ from overridden 'get' method
************* Module lms.djangoapps.grades.api.serializers
lms/djangoapps/grades/api/serializers.py:19: [W0221(arguments-differ), GradingPolicySerializer.to_representation] Parameters differ from overridden 'to_representation' method
************* Module lms.djangoapps.grades.api.tests.test_views
lms/djangoapps/grades/api/tests/test_views.py:326: [I0021(useless-suppression), ] Useless suppression of 'missing-docstring'
lms/djangoapps/grades/api/tests/test_views.py:319: [I0021(useless-suppression), ] Useless suppression of 'missing-docstring'
lms/djangoapps/grades/api/tests/test_views.py:257: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:196: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:198: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:233: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:235: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:172: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:173: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:270: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:271: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:304: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:184: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:214: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:183: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:216: [I0021(useless-suppression), ] Useless suppression of 'no-member'
lms/djangoapps/grades/api/tests/test_views.py:286: [I0021(useless-suppression), ] Useless suppression of 'no-member'
```

kabirkhan commented on lms/djangoapps/grades/api/v1/views.py:
```diff
@@ -0,0 +1,202 @@
+        * course_id: (required) A string representation of a Course ID.
+        * username:  (optional) A string representation of a user's username.
+    **GET Response Values**
+        If the request f    or information about the course grade
```
@nasthagiri so you want me to manually squash commits and then rebase?
I've just been pulling the latest master changes each time I make new commits. 

Generally we use the squash PR functionality inside github when merging pull requests to achieve this effect. At this point it's a bit non-trivial to squash all these commits and then rebase without using a different branch.

Let me know if you have any guidance.

nasthagiri replied:
I see. Here, we generally do this via the command line:

```
# to squash all 21 commits
git rebase -i HEAD~21
# in the editor, replace all "pick" with "s" - except for the 1st commit

# to rebase with master
git fetch; git rebase origin/master

# to push back changes (you need a force-push since you've rebased)
git push -f
```

@nedbat Do you have alternative advice for @kabirkhan?

kabirkhan replied:
@nasthagiri @nedbat the end effect should be the same either way: a single squashed commit merged to master.

nasthagiri replied:
@kabirkhan I would be the one merging this PR and that option is not available to me in this repo:
![screen shot 2018-05-02 at 5 17 50 pm](https://user-images.githubusercontent.com/6772826/39549691-c9d9e2c4-4e2c-11e8-8956-3e2eba5849ca.png)

kabirkhan replied:
@nasthagiri I believe I've done this properly.

nasthagiri replied:
Ok. great!  As soon as the build is green, I'll merge.

edx-webhook commented:
@kabirkhan 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/17602&entry.1484655318=https://github.com/kabirkhan&entry.752974735=2018-03-01+22:42&entry.1917517419=2018-05-03+19:36&entry.2133058324=Yes) so we can improve your experience in the future.

nasthagiri commented:
🎉 Congratulations @kabirkhan and Microsoft Worldwide Learning!!! 🎉

nasthagiri commented:
FYI @edx/educator-neem and @edx/educator-escalation on v1 of the Grades API getting merged, with support for bulk read.
